pkg/deviceplugin/topology.go

```diff
@@ -0,0 +1,254 @@
+// Copyright 2019 Intel Corporation. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Most of this code was copied from CRI Resource Manager:
+// https://github.com/intel/cri-resource-manager/blob/v0.2.0/pkg/sysfs/topology.go
+
+package deviceplugin
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"syscall"
+
+	"github.com/pkg/errors"
+	"golang.org/x/sys/unix"
+
+	pluginapi "k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+)
+
+// to mock in tests
+var (
+	mockRoot = ""
+)
+
+const (
+	// ProviderKubelet is a constant to distinguish that topology hint comes
+	// from parameters passed to CRI create/update requests from Kubelet
+	ProviderKubelet = "kubelet"
+)
+
+// TopologyHint represents various hints that can be detected from sysfs for the device
+type TopologyHint struct {
+	Provider string
+	CPUs     string
+	NUMAs    string
+	Sockets  string
+}
+
+// TopologyHints represents set of hints collected from multiple providers
+type TopologyHints map[string]TopologyHint
+
+// GetTopologyInfo returns topology information for the list of device nodes
+func GetTopologyInfo(devs []string) (*pluginapi.TopologyInfo, error) {
+	var result pluginapi.TopologyInfo
+	nodeIDs := map[int64]bool{}
```

Comment:
it become more ugly in the assignments (like `struct{}{}`), there are other disadvantages that can be seen here: 
https://dave.cheney.net/2014/03/25/the-empty-struct#comment-2826

Reply:
I wouldn't trust boolean-based sets because for example `len(m2)` can be misleading. In the long run sets implemented with empty structs are less error prone.